renderdoc/driver/d3d12/d3d12_shaderdebug.cpp

```diff
@@ -117,10 +118,211 @@ bool D3D12DebugAPIWrapper::CalculateSampleGather(
+        {
+          UINT shaderReg = param.Descriptor.ShaderRegister;
+          ID3D12Resource *pResource = rm->GetCurrentAs<ID3D12Resource>(element.id);
+          D3D12_RESOURCE_DESC resDesc = pResource->GetDesc();
+
+          ShaderDebug::FillViewFmt(DXGI_FORMAT_R32_UINT, global.srvs[shaderReg].format);
+          global.srvs[shaderReg].firstElement = (uint32_t)(element.offset / sizeof(uint32_t));
+          global.srvs[shaderReg].numElements =
+              (uint32_t)((resDesc.Width - element.offset) / sizeof(uint32_t));
+
+          if(resDesc.Dimension == D3D12_RESOURCE_DIMENSION_BUFFER)
+            GetBufferData(pResource, 0, 0, global.srvs[shaderReg].data);
+        }
+        else if(param.ParameterType == D3D12_ROOT_PARAMETER_TYPE_UAV && element.type == eRootUAV)
+        {
+          UINT shaderReg = param.Descriptor.ShaderRegister;
+          ID3D12Resource *pResource = rm->GetCurrentAs<ID3D12Resource>(element.id);
+          D3D12_RESOURCE_DESC resDesc = pResource->GetDesc();
+
+          // TODO: Need to fetch counter resource if applicable
+
+          ShaderDebug::FillViewFmt(DXGI_FORMAT_R32_UINT, global.uavs[shaderReg].format);
+          global.uavs[shaderReg].firstElement = (uint32_t)(element.offset / sizeof(uint32_t));
+          global.uavs[shaderReg].numElements =
+              (uint32_t)((resDesc.Width - element.offset) / sizeof(uint32_t));
+
+          if(resDesc.Dimension == D3D12_RESOURCE_DIMENSION_BUFFER)
+            GetBufferData(pResource, 0, 0, global.uavs[shaderReg].data);
+
+          // TODO: Need to handle case where UAV isn't a buffer
+        }
+        else if(param.ParameterType == D3D12_ROOT_PARAMETER_TYPE_DESCRIPTOR_TABLE &&
+                element.type == eRootTable)
+        {
+          UINT prevTableOffset = 0;
+          WrappedID3D12DescriptorHeap *heap =
+              rm->GetCurrentAs<WrappedID3D12DescriptorHeap>(element.id);
+
+          size_t numRanges = param.ranges.size();
+          for(size_t r = 0; r < numRanges; ++r)
+          {
+            const D3D12_DESCRIPTOR_RANGE1 &range = param.ranges[r];
+
+            UINT offset = range.OffsetInDescriptorsFromTableStart;
+            if(range.OffsetInDescriptorsFromTableStart == D3D12_DESCRIPTOR_RANGE_OFFSET_APPEND)
+              offset = prevTableOffset;
+
+            D3D12Descriptor *desc = (D3D12Descriptor *)heap->GetCPUDescriptorHandleForHeapStart().ptr;
+            desc += element.offset;
+            desc += offset;
+
+            UINT numDescriptors = range.NumDescriptors;
+            if(numDescriptors == UINT_MAX)
+            {
+              // Find out how many descriptors are left after
+              UINT availDescriptors = heap->GetNumDescriptors() - offset - (UINT)element.offset;
+
+              // TODO: Should we look up the bind point in the D3D12 state to try to get
+              // a better guess at the number of descriptors?
+
+              numDescriptors = RDCMIN(availDescriptors, 128U);
```

Comment:
I'm not sure I understand this min with 128.

About the TODO, while this isn't necessary for correctness I think anything we can do to reduce the number of descriptors processed is a good thing. Otherwise this is a lurking problem that can potentially explode whenever someone is debugging a bindless-style shader with millions of descriptors bound (which sadly is reasonably common in D3D12 from what I've seen).

Reply:
The 128 is mimicking logic in D3D12Replay::FillRegisterSpaces, where a matching shader binding isn't found.

Reply:
Ah I see. In that case this probably shouldn't be present until you are searching the bindings, otherwise you'd be removing descriptors that might be used. That case was to reduce the amount of redundant descriptors to pass via the pipeline state when there's no binding at all (only visible if you enable 'show unused/disabled'). For this case if you don't have a matching binding you can skip processing the descriptors entirely because it's impossible for the shader to use it.

Reply:
Thanks for the explanation. I removed the min and left the TODO